rpm/files.go

```diff
@@ -0,0 +1,139 @@
+	refs map[string]int
+	mu   sync.Mutex
+	sf   singleflight.Group
+}
+
+var fc = &filesCache{
+	c:    map[string]map[string]struct{}{},
+	refs: map[string]int{},
+}
+
+// GC decrements the reference counter and deletes the [claircore.Layer]'s
+// entry from the cache map and the reference counter once the references
+// are 0 (or less).
+func (fc *filesCache) gc(key string) {
+	fc.mu.Lock()
+	defer fc.mu.Unlock()
+	fc.refs[key]--
+	if fc.refs[key] <= 0 {
+		delete(fc.c, key)
+		delete(fc.refs, key)
+	}
+}
+
+// Get increments the reference counter for the key and looks up the key
+// in the cache. It will subsequently start a goroutine to track when the
+// calling context is done, calling the gc method.
+func (fc *filesCache) get(ctx context.Context, key string) (map[string]struct{}, bool) {
+	fc.mu.Lock()
+	defer fc.mu.Unlock()
+	fc.refs[key]++
+	f, ok := fc.c[key]
+	go func() {
+		<-ctx.Done()
+		fc.gc(key)
+	}()
+	return f, ok
+}
+
+// Set sets the files for a particular key and deals with the locking.
+func (fc *filesCache) set(key string, files map[string]struct{}) {
+	fc.mu.Lock()
+	defer fc.mu.Unlock()
+	fc.c[key] = files
+}
+
+var errNoDBFound = errors.New("no RPM DB found")
+
+// GetFiles looks up RPM files that exist in the RPM database using the
+// filesFromDB function and memorizes the result to avoid repeated work
+// for the same [claircore.Layer].
+func (fc *filesCache) getFiles(ctx context.Context, layer *claircore.Layer) (map[string]struct{}, error) {
+	if fc == nil {
+		panic("programmer error: filesCache nil")
+	}
+	key := layer.Hash.String()
+	if files, ok := fc.get(ctx, key); ok {
+		return files, nil
+	}
+
+	files := map[string]struct{}{}
+	ch := fc.sf.DoChan(key, func() (interface{}, error) {
+		sys, err := layer.FS()
+		if err != nil {
+			return nil, fmt.Errorf("rpm: unable to open layer: %w", err)
+		}
+
+		found := make([]foundDB, 0)
+		if err := fs.WalkDir(sys, ".", findDBs(ctx, &found, sys)); err != nil {
+			return nil, fmt.Errorf("rpm: error walking fs: %w", err)
+		}
+		if len(found) == 0 {
+			return nil, errNoDBFound
+		}
+
+		done := map[string]struct{}{}
+		zlog.Debug(ctx).Int("count", len(found)).Msg("found possible databases")
+		for _, db := range found {
+			ctx := zlog.ContextWithValues(ctx, "db", db.String())
+			zlog.Debug(ctx).Msg("examining database")
+			if _, ok := done[db.Path]; ok {
+				zlog.Debug(ctx).Msg("already seen, skipping")
+				continue
+			}
+			done[db.Path] = struct{}{}
+			fs, err := getDBObjects(ctx, sys, db, filesFromDB)
+			if err != nil {
+				return nil, fmt.Errorf("rpm: error getting native DBs: %w", err)
+			}
+			for _, f := range fs {
+				files[f.Path] = struct{}{}
+			}
+		}
+		return files, nil
+	})
+	select {
+	case <-ctx.Done():
+		return nil, context.Cause(ctx)
+	case res := <-ch:
+		switch {
+		case res.Err == nil:
+			files = res.Val.(map[string]struct{})
+		case errors.Is(res.Err, errNoDBFound):
+		default:
+			return nil, res.Err
+		}
+	}
+
+	fc.set(key, files)
```

Comment:
I think we should call this from within the `DoChan`. Otherwise, we may see a bunch of calls to this for the same layer. We can avoid it by adding the files from within the `DoChan`, so it'll only be called once per layer